syntax = "proto3";
package imgui;

enum MsgType
{
    CS_NONE    = 0;
    SC_FRAME   = 1; //发送帧数据
    SC_TEXTURE = 2; //发送贴图
    CS_IO      = 3; //发送操作
}

message Vec2
{
    float x = 1;
    float y = 2;
}

message Vec4
{
    float x = 1;
    float y = 2;
    float z = 3;
    float w = 4;
}

message DrawCmd
{
    Vec4   ClipRect  = 1;
    uint64 TextureId = 2;
    uint32 VtxOffset = 3;
    uint32 IdxOffset = 4;
    uint32 ElemCount = 5;
}

message DrawVert
{
    Vec2   pos = 1;
    Vec2   uv  = 2;
    uint32 col = 3;
}

message DrawList
{
    repeated DrawCmd    CmdBuffer = 1; 
    repeated uint32     IdxBuffer = 2; 
    repeated DrawVert   VtxBuffer = 3; 
    uint32              Flags     = 4; 
}

message FrameData
{

    bool              Valid            = 1;
    repeated DrawList CmdLists         = 2;
    int32             CmdListsCount    = 3;
    int32             TotalIdxCount    = 4;
    int32             TotalVtxCount    = 5;
    Vec2              DisplayPos       = 6;
    Vec2              DisplaySize      = 7;
    Vec2              FramebufferScale = 8;
}

message IOData
{
    Vec2            MousePos  = 1;
    repeated uint32 MouseDown = 2;
}

message TextureData
{
    uint32 tex_id = 1;
    int32  width  = 2;
    int32  height = 3;
    bytes  pixels = 5;
}